File: src/javabot/operations/DaysUntilOperation.java
package javabot.operations;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;

import javabot.BotEvent;
import javabot.Message;

import com.rickyclarkson.java.util.TypeSafeList;

/**
 * @author littlezoper
 */
public class DaysUntilOperation implements BotOperation
{
	/**
		@see BotOperation#handleMessage(BotEvent)
	*/
	public List handleMessage(BotEvent event)
	{
		List messages=new TypeSafeList(new ArrayList(),Message.class);
		
		String message = event.getMessage().toLowerCase();
		String sender = event.getSender();
		if (!message.startsWith("days until "))
			return messages;
			
		message = message.substring("days until ".length());

		Calendar calendar=Calendar.getInstance();
		SimpleDateFormat sdf = new SimpleDateFormat();
		Date d = null;

		String formats[] = {"yyyy/MM/dd", "MMM d, ''yy", "d MMM yyyy",
			"MMM d, yyyy", "MMM d, ''yy"};

		int i = 0;
		while ((i < formats.length) && (d == null)) {
			sdf.applyPattern(formats[i]);
			try
			{
				d = sdf.parse (message);
			} catch (ParseException e) {
				// I think we just want to ignore this...
			}

			i++;
		}


		if (d == null) {
			messages.add (new Message(event.getChannel(),
				sender + ":  you might want to consider putting the date in a proper format...",
				false));

			return messages;
		}

		calendar.setTime(d);
		Calendar today=Calendar.getInstance();
		today.set(Calendar.HOUR_OF_DAY, 0);
		today.set(Calendar.MINUTE, 0);
		today.set(Calendar.SECOND, 0);
		today.set(Calendar.MILLISECOND, 0);
		
		long millis=calendar.getTimeInMillis()-today.getTimeInMillis();
		
		double days = millis /= 86400000;
		
		messages.add (new Message (event.getChannel(),
				sender + ":  there are "+(int)days+" days until " + message + ".",
				false));
		
		return messages;
	}
}

File: src/main/kotlin/javabot/operations/DaysUntilOperation.kt
package javabot.operations

import com.antwerkz.sofia.Sofia
import com.google.inject.Inject
import java.time.Duration
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.Date
import java.util.Locale
import javabot.Javabot
import javabot.Message
import javabot.dao.AdminDao

class DaysUntilOperation @Inject constructor(bot: Javabot, adminDao: AdminDao) :
    BotOperation(bot, adminDao) {
    override fun handleMessage(event: Message): List<Message> {
        val responses = arrayListOf<Message>()
        var message = event.value.lowercase(Locale.getDefault())
        if (message.startsWith("days until ")) {
            val sender = event.user
            message = message.substring("days until ".length)
            var d: LocalDateTime? = null
            val formats =
                arrayOf(
                    DateTimeFormatter.ofPattern("yyyy/MM/dd"),
                    DateTimeFormatter.ofPattern("MMM d, ''yy"),
                    DateTimeFormatter.ofPattern("d MMM yyyy"),
                    DateTimeFormatter.ofPattern("MMM d, yyyy"),
                    DateTimeFormatter.ofPattern("MMM d, ''yy")
                )
            var i = 0
            while (i < formats.size && d == null) {
                try {
                    d = LocalDateTime.parse(message, formats[i])
                    calcTime(responses, event, d)
                } catch (e: IllegalArgumentException) {
                    // I think we just want to ignore this...
                }

                i++
            }
            if (d == null) {
                responses.add(Message(event, Sofia.invalidDateFormat(sender.nick)))
            }
        }
        return responses
    }

    private fun calcTime(responses: MutableList<Message>, event: Message, d: LocalDateTime) {
        val days = Duration.between(d, LocalDateTime.now().withHour(0)).toDays()
        val l = d.toLocalDate().toEpochDay()
        responses.add(Message(event, Sofia.daysUntil(event.user.nick, days, Date(l))))
    }
}
